lisa/runners/lisa_runner.py

```diff
@@ -20,298 +21,384 @@
+                return self._generate_task(
+                    task_method=self._initialize_environment_task,
+                    environment=environment,
+                    test_results=initialization_results,
                 )
-            finally:
-                if environment and environment.status in [
-                    EnvironmentStatus.Deployed,
-                    EnvironmentStatus.Connected,
-                ]:
-                    platform.delete_environment(environment)
-
-        # not run as there is no fit environment.
-        for case in self._get_can_run_results(can_run_results):
-            reasons = "no available environment"
-            if case.check_results and case.check_results.reasons:
-                reasons = f"{reasons}: {case.check_results.reasons}"
 
-            case.set_status(TestStatus.SKIPPED, reasons)
+        # run on connected environment
+        can_run_results = [x for x in can_run_results if x.can_run]
+        if environment.status == EnvironmentStatus.Connected and can_run_results:
+            selected_test_results = self._get_test_results_to_run(
+                test_results=test_results, environment=environment
+            )
+            if selected_test_results:
+                return self._generate_task(
+                    task_method=self._run_test_task,
+                    environment=environment,
+                    test_results=selected_test_results,
+                )
 
-        self.status = ActionStatus.SUCCESS
+        return None
 
-        # for UT testability
-        self._latest_platform = platform
+    def _delete_unused_environments(self) -> Optional[Callable[[], List[TestResult]]]:
+        available_environments = self._sort_environments(self.environments)
+        # check deleteable environments
+        for environment in available_environments:
+            if environment.is_in_use:
+                continue
 
-        return test_results
+            can_run_results = self._get_runnable_test_results(
+                self.test_results, environment=environment
+            )
+            if not can_run_results:
+                # no more test need this environment, delete it.
+                self._log.debug(
+                    f"generating delete environment task on '{environment.name}'"
+                )
+                return self._generate_task(
+                    task_method=self._delete_environment_task,
+                    environment=environment,
+                    test_results=[],
+                )
+        return None
 
     def _prepare_environments(
         self,
         platform: Platform,
-        candidate_environments: Environments,
         test_results: List[TestResult],
-    ) -> List[Environment]:
+    ) -> List[TestResult]:
+        if self._is_prepared:
+            return []
+
+        runbook_environments = load_environments(self._runbook.environment)
+        if not runbook_environments:
```

Comment:
It will always return an Enviroments object, how can that be evaluated as false?

Reply:
`load_environments` loads environments from runbook, but the environment in runbook is optional. the next condition will create environments from test case requirements, if there is no environment defined in runbook.

Reply:
else:
        environments = Environments()
this is what I'm talking about--that can't be evaluated as false, can it?

Reply:
The Environments is a dict, and it may be empty.

Reply:
I see, missed one more step in my look up. All good, disregard :)